Web Driver: [Cocoa] Viewport snapshots do not correctly represent CSS transforms

patrickangle opened the pull request:
#### dc2acd22d6be4307f9ac8b57b2b26a7ac3437dfe
<pre>
Web Driver: [Cocoa] Viewport snapshots do not correctly represent CSS transforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=242214">https://bugs.webkit.org/show_bug.cgi?id=242214</a>
rdar://33925578

Reviewed by Devin Rousso.

In order to get accurate snapshots of CSS Transforms, we have to capture the snapshot from the UIProcess. By doing so we
get the actual pixels that would be/are drawn onto the screen at the expense of an inability to capture a snapshot of
any pixels that are not within the visible viewport at the time of capture. For that reason this patch falls back to the
existing WebProcess snapshotting for element snapshots, since an element may not be entirely within the viewport. This
change to the viewport snapshot behavior progresses a number of WPT tests that previously failed including but not
limited to:
- css/css-transforms/perspective-origin-001.html (through 006)
- css/css-transforms/perspective-origin-001.html
- css/css-transforms/perspective-origin-002.html
- css/css-transforms/perspective-origin-003.html
- css/css-transforms/perspective-origin-004.html
- css/css-transforms/perspective-origin-005.html
- css/css-transforms/perspective-origin-006.html
- css/css-transforms/perspective-origin-x.html
- css/css-transforms/perspective-origin-xy.html
- css/css-transforms/perspective-translateZ-0.html
- css/css-transforms/perspective-translateZ-negative.html
- css/css-transforms/perspective-translateZ-positive.html
- css/css-transforms/text-perspective-001.html

There is a minor annoyance for viewport snapshots on macOS which is that the bottom left and right corners will appear
rounded, which is an unfortunate side effect of taking the snapshot of the window (taking a snapshot of the view itself
results in an empty image). I believe however this is a worthwhile tradeoff for passing all these tests, and likely many
more. iOS in not affected in this way.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::getBase64EncodedPNGData):
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::takeViewSnapshot):
- Match the behavior of -[WKWebViewIOS _takeViewSnapshot] and return a non-volatile surface, otherwise the bits are
often missing by the time we are trying to base64 encode them.

Canonical link: <a href="https://commits.webkit.org/252201@main">https://commits.webkit.org/252201@main</a>
</pre>

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2307,7 +2307,24 @@ void WebAutomationSession::takeScreenshot(const Inspector::Protocol::Automation:
+    // snapshot of the entire viewport without the window's rounded corners being excluded. So we trade off those corner
+    // pixels for accurate pixels in the rest of the viewport which help us verify features like CSS transforms are
+    // actually behaving correctly.
+    if (!nodeHandle.isEmpty()) {
```
I wonder if maybe we could try to be a teensy bit clever here and check if the place the node would be rendered at is entirely within the viewport, using `takeViewSnapshot` if so 🤔

patrickangle replied:
I had that thought, yeah... For now I want to keep the behavior predictable such that the same call to webdriver will produce predictable results regardless of where an element is in relation to the viewport.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2307,7 +2307,24 @@ void WebAutomationSession::takeScreenshot(const Inspector::Protocol::Automation:
+#endif
+
     Function<void(WebPageProxy&, std::optional<WebCore::IntRect>&&, Ref<TakeScreenshotCallback>&&)> takeViewSnapsot = [](WebPageProxy& page, std::optional<WebCore::IntRect>&& rect, Ref<TakeScreenshotCallback>&& callback) {
         page.callAfterNextPresentationUpdate([page = Ref { page }, rect = WTFMove(rect), callback = WTFMove(callback)](CallbackBase::Error error) mutable {
```
I think this new path will have slightly different timing behavior.  Is that problematic?

patrickangle replied:
I don't believe so, no. It's hard to draw a comparison between the two paths because the other path will actually paint fresh into a bitmap for us which means it may also not give us the "current frame" as visible and may be representative of the next frame.

The fact that GTK has been going down this path for a bit adds to my confidence.

patrickangle committed:
Web Driver: [Cocoa] Viewport snapshots do not correctly represent CSS transforms
https://bugs.webkit.org/show_bug.cgi?id=242214
rdar://33925578

Reviewed by Devin Rousso.

In order to get accurate snapshots of CSS Transforms, we have to capture the snapshot from the UIProcess. By doing so we
get the actual pixels that would be/are drawn onto the screen at the expense of an inability to capture a snapshot of
any pixels that are not within the visible viewport at the time of capture. For that reason this patch falls back to the
existing WebProcess snapshotting for element snapshots, since an element may not be entirely within the viewport. This
change to the viewport snapshot behavior progresses a number of WPT tests that previously failed including but not
limited to:
- css/css-transforms/perspective-origin-001.html (through 006)
- css/css-transforms/perspective-origin-001.html
- css/css-transforms/perspective-origin-002.html
- css/css-transforms/perspective-origin-003.html
- css/css-transforms/perspective-origin-004.html
- css/css-transforms/perspective-origin-005.html
- css/css-transforms/perspective-origin-006.html
- css/css-transforms/perspective-origin-x.html
- css/css-transforms/perspective-origin-xy.html
- css/css-transforms/perspective-translateZ-0.html
- css/css-transforms/perspective-translateZ-negative.html
- css/css-transforms/perspective-translateZ-positive.html
- css/css-transforms/text-perspective-001.html

There is a minor annoyance for viewport snapshots on macOS which is that the bottom left and right corners will appear
rounded, which is an unfortunate side effect of taking the snapshot of the window (taking a snapshot of the view itself
results in an empty image). I believe however this is a worthwhile tradeoff for passing all these tests, and likely many
more. iOS in not affected in this way.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::getBase64EncodedPNGData):
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::takeViewSnapshot):
- Match the behavior of -[WKWebViewIOS _takeViewSnapshot] and return a non-volatile surface, otherwise the bits are
often missing by the time we are trying to base64 encode them.

Canonical link: https://commits.webkit.org/252201@main
